File: HoTT-MdlViewer-JavaFX/src/main/java/gde/mdl/ui/Controller.java
package gde.mdl.ui;

import java.io.File;
import java.io.IOException;
import java.util.List;
import java.util.Optional;
import java.util.prefs.BackingStoreException;
import java.util.prefs.Preferences;
import java.util.stream.Collectors;

import javax.xml.bind.JAXBException;

import org.apache.commons.io.FilenameUtils;

import com.itextpdf.text.DocumentException;

import de.treichels.hott.HoTTDecoder;
import gde.mdl.messages.Messages;
import gde.mdl.ui.background.RefreshService;
import gde.model.BaseModel;
import gde.model.voice.Announcements;
import gde.model.voice.VoiceFile;
import gde.report.ReportException;
import javafx.application.Application;
import javafx.application.Platform;
import javafx.beans.binding.Bindings;
import javafx.beans.property.ObjectProperty;
import javafx.beans.property.ReadOnlyBooleanProperty;
import javafx.beans.property.SimpleObjectProperty;
import javafx.concurrent.Task;
import javafx.fxml.FXML;
import javafx.fxml.FXMLLoader;
import javafx.scene.Cursor;
import javafx.scene.Scene;
import javafx.scene.control.ContextMenu;
import javafx.scene.control.MenuItem;
import javafx.scene.image.Image;
import javafx.scene.input.MouseButton;
import javafx.scene.input.MouseEvent;
import javafx.scene.layout.BorderPane;
import javafx.scene.layout.Region;
import javafx.scene.layout.StackPane;
import javafx.scene.web.WebView;
import javafx.stage.FileChooser;
import javafx.stage.FileChooser.ExtensionFilter;
import javafx.stage.Stage;

public class Controller extends Application {
    private static final String LAST_LOAD_DIR = "lastLoadDir"; //$NON-NLS-1$
    private static final String LAST_SAVE_DIR = "lastSaveDir"; //$NON-NLS-1$
    static final Preferences PREFS = Preferences.userNodeForPackage(Controller.class);
    static Stage STAGE;

    @FXML
    private BorderPane borderPane;

    @FXML
    private StackPane stackPane;

    @FXML
    private ContextMenu contextMenu;

    @FXML
    private Region overlay;

    @FXML
    private MenuItem refresh1;

    @FXML
    private MenuItem refresh2;

    @FXML
    private MenuItem save1;

    @FXML
    private MenuItem save2;

    @FXML
    private MenuItem sysVDF1;

    @FXML
    private MenuItem sysVDF2;

    @FXML
    private MenuItem userVDF1;

    @FXML
    private MenuItem userVDF2;

    @FXML
    private WebView webview;

    private Scene scene;
    private RefreshService refreshService;
    private final ObjectProperty<Model> modelProperty = new SimpleObjectProperty<>(null);

    private void disableUI(final ReadOnlyBooleanProperty readOnlyBooleanProperty) {
        scene.cursorProperty().bind(Bindings.when(readOnlyBooleanProperty).then(Cursor.WAIT).otherwise(Cursor.DEFAULT));
        overlay.visibleProperty().bind(readOnlyBooleanProperty);
        webview.visibleProperty().bind(readOnlyBooleanProperty.not());
        readOnlyBooleanProperty.addListener((p, o, n) -> borderPane.setDisable(n));
    }

    @FXML
    public void initialize() throws IOException {
        save1.disableProperty().bind(modelProperty.isNull());
        save2.disableProperty().bind(modelProperty.isNull());
        refresh1.disableProperty().bind(modelProperty.isNull());
        refresh2.disableProperty().bind(modelProperty.isNull());
        sysVDF1.disableProperty().bind(modelProperty.isNull());
        userVDF1.disableProperty().bind(modelProperty.isNull());
        sysVDF2.disableProperty().bind(modelProperty.isNull());
        userVDF2.disableProperty().bind(modelProperty.isNull());

        refreshService = new RefreshService(webview);

        final LoadingIndicator indicator = new LoadingIndicator();
        indicator.visibleProperty().bind(overlay.visibleProperty());

        stackPane.getChildren().add(indicator);

        // FIXME: FXMLLoader does not set contextMenuEnabled correctly.
        // Therefore, we have to disable it manually.
        webview.setContextMenuEnabled(false);
    }

    @FXML
    public void onExit() {
        Platform.exit();
    }

    @FXML
    public void onLoadFromFile() {
        final FileChooser chooser = new FileChooser();
        chooser.setTitle(Messages.getString("LoadFromFile"));
        final File dir = new File(PREFS.get(LAST_LOAD_DIR, System.getProperty(Launcher.MDL_DIR)));
        if (dir.exists() && dir.isDirectory()) chooser.setInitialDirectory(dir);
        chooser.getExtensionFilters().add(new ExtensionFilter(Messages.getString("MdlFileDescription"), "*.mdl"));

        final File file = chooser.showOpenDialog(STAGE);
        if (file != null) {
            PREFS.put(LAST_LOAD_DIR, file.getParentFile().getAbsolutePath());
            try {
                modelProperty.set(Model.load(file));
            } catch (final IOException e) {
                ExceptionDialog.show(e);
            }
            onRefresh();
        }
    }

    @FXML
    public void onLoadFromMemory() {
        final SelectFromTransmitter dialog = new SelectFromMemory();
        final Optional<Task<Model>> result = dialog.showAndWait();
        if (result.isPresent()) {
            final Task<Model> task = result.get();
            disableUI(task.runningProperty());
            task.setOnSucceeded(e -> {
                modelProperty.set(task.getValue());
                onRefresh();
            });
        }
    }

    @FXML
    public void onLoadFromSdCard() {
        final SelectFromSdCard dialog = new SelectFromSdCard();
        final Optional<Task<Model>> result = dialog.showAndWait();
        if (result.isPresent()) {
            final Task<Model> task = result.get();
            disableUI(task.runningProperty());
            task.setOnSucceeded(e -> {
                modelProperty.set(task.getValue());
                onRefresh();
            });
        }
    }

    @FXML
    public void onLoadSystemVDF() throws IOException, BackingStoreException {
        final FileChooser chooser = new FileChooser();
        chooser.setTitle(Messages.getString("LoadSystemVDF"));
        final File dir = new File(PREFS.get(LAST_LOAD_DIR, System.getProperty(Launcher.MDL_DIR)));
        if (dir.exists() && dir.isDirectory()) chooser.setInitialDirectory(dir);
        chooser.getExtensionFilters().add(new ExtensionFilter(Messages.getString("VdfFileDescription"), "*.vdf"));

        final File vdf = chooser.showOpenDialog(STAGE);
        if (vdf != null) {
            PREFS.put(LAST_LOAD_DIR, vdf.getParentFile().getAbsolutePath());
            final BaseModel model = modelProperty.get().decode();
            final VoiceFile vdfFile = HoTTDecoder.decodeVDF(vdf);
            final List<String> names = vdfFile.getVoiceList().stream().map(v -> v.getName()).collect(Collectors.toList());
            Announcements.saveSystemPrefs(model.getTransmitterType(), names);
            onRefresh();
        }
    }

    @FXML
    public void onLoadUserVDF() throws IOException, BackingStoreException {
        final FileChooser chooser = new FileChooser();
        chooser.setTitle(Messages.getString("LoadUserVDF"));
        final File dir = new File(PREFS.get(LAST_LOAD_DIR, System.getProperty(Launcher.MDL_DIR)));
        if (dir.exists() && dir.isDirectory()) chooser.setInitialDirectory(dir);
        chooser.getExtensionFilters().add(new ExtensionFilter(Messages.getString("VdfFileDescription"), "*.vdf"));

        final File vdf = chooser.showOpenDialog(STAGE);
        if (vdf != null) {
            PREFS.put(LAST_LOAD_DIR, vdf.getParentFile().getAbsolutePath());
            final BaseModel model = modelProperty.get().decode();
            final VoiceFile vdfFile = HoTTDecoder.decodeVDF(vdf);
            final List<String> names = vdfFile.getVoiceList().stream().map(v -> v.getName()).collect(Collectors.toList());
            Announcements.saveUserPrefs(model.getTransmitterType(), names);
            onRefresh();
        }
    }

    @FXML
    public void onMouseClicked(final MouseEvent e) {
        if (e.getButton() == MouseButton.SECONDARY)
            contextMenu.show(webview, e.getScreenX(), e.getScreenY());
        else
            contextMenu.hide();
    }

    @FXML
    public void onRefresh() {
        if (modelProperty.isNotNull().get()) {
            refreshService.start(modelProperty.get());
            disableUI(refreshService.runningProperty());
        }
    }

    @FXML
    public void onSave() {
        final FileChooser chooser = new FileChooser();
        chooser.setTitle(Messages.getString("Save"));
        final File dir = new File(PREFS.get(LAST_SAVE_DIR, System.getProperty(Launcher.MDL_DIR)));
        if (dir.exists() && dir.isDirectory()) chooser.setInitialDirectory(dir);
        chooser.setInitialFileName(modelProperty.get().getFileName());
        chooser.getExtensionFilters().add(new ExtensionFilter(Messages.getString("SimpleGUI.HTML"), "*.html"));
        chooser.getExtensionFilters().add(new ExtensionFilter(Messages.getString("SimpleGUI.PDF"), "*.pdf"));
        chooser.getExtensionFilters().add(new ExtensionFilter(Messages.getString("SimpleGUI.XML"), "*.xml"));
        chooser.getExtensionFilters().add(new ExtensionFilter(Messages.getString("MdlFileDescription"), "*.mdl"));
        chooser.getExtensionFilters().add(new ExtensionFilter(Messages.getString("TextDump"), "*.txt"));

        final File fileToSave = chooser.showSaveDialog(STAGE);
        if (fileToSave != null) {
            PREFS.put(LAST_SAVE_DIR, fileToSave.getParentFile().getAbsolutePath());

            final String extension = FilenameUtils.getExtension(fileToSave.getName()).toLowerCase();
            final Model model = modelProperty.get();

            final Runnable r = () -> {
                try {
                    switch (extension) {
                    case "html":
                        model.saveHtml(fileToSave);
                        break;

                    case "pdf":
                        model.savePdf(fileToSave);
                        break;

                    case "xml":
                        model.saveXml(fileToSave);
                        break;

                    case "mdl":
                        model.saveMdl(fileToSave);
                        break;

                    case "txt":
                        model.saveTxt(fileToSave);
                    }
                } catch (ReportException | IOException | DocumentException | JAXBException e) {
                    ExceptionDialog.show(e);
                }
            };

            new Thread(r).start();
        }
    }

    @Override
    public void start(final Stage primaryStage) throws Exception {
        final String title = Messages.getString("SimpleGUI.Title", System.getProperty(Launcher.PROGRAM_VERSION));
        final FXMLLoader loader = new FXMLLoader(getClass().getResource("MdlViewer.fxml"), Messages.getResourceBundle());
        loader.setController(this);
        scene = new Scene(loader.load());
        final Image icon = new Image(getClass().getResource("/icon.png").toString());
        STAGE = primaryStage;

        primaryStage.getIcons().add(icon);
        primaryStage.setTitle(title);
        primaryStage.setScene(scene);
        primaryStage.show();
    }
}

File: HoTT-MdlViewer-JavaFX/src/main/java/de/treichels/hott/mdlviewer/javafx/Controller.kt
/**
 * HoTT Transmitter Config Copyright (C) 2013 Oliver Treichel
 *
 * This program is free software: you can redistribute it and/or modify it under the terms of the GNU General Public License as published by the Free Software
 * Foundation, either version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR
 * A PARTICULAR PURPOSE. See the GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along with this program. If not, see <http:></http:>//www.gnu.org/licenses/>.
 */
package de.treichels.hott.mdlviewer.javafx

import de.treichels.decoder.HoTTDecoder
import de.treichels.hott.model.voice.Announcements
import javafx.application.Platform
import javafx.beans.binding.Bindings
import javafx.beans.property.ReadOnlyBooleanProperty
import javafx.beans.property.SimpleObjectProperty
import javafx.scene.Cursor
import javafx.scene.control.ContextMenu
import javafx.scene.image.Image
import javafx.scene.input.MouseButton
import javafx.scene.input.MouseEvent
import javafx.scene.web.WebView
import javafx.stage.FileChooser
import javafx.stage.FileChooser.ExtensionFilter
import kotlinx.coroutines.experimental.async
import org.apache.commons.io.FilenameUtils
import tornadofx.*
import java.io.File
import java.io.FileWriter

private const val LAST_LOAD_DIR = "lastLoadDir"
private const val LAST_SAVE_DIR = "lastSaveDir"

class Controller : View() {
    private val selectFromMemory by inject<SelectFromMemory>()
    private val selectFromSDCard by inject<SelectFromSdCard>()
    private var contextMenu by singleAssign<ContextMenu>()
    private var webView by singleAssign<WebView>()
    private val modelProperty = SimpleObjectProperty<Model?>(null)
    private var model by modelProperty

    override val root = borderpane {
        prefWidth = 800.0
        prefHeight = 600.0

        top {
            menubar {
                menu(messages["file"]) {
                    item(messages["load_from_file"], "Shortcut+F1") {
                        action { loadFromFile() }
                    }
                    item(messages["load_from_memory"], "Shortcut+F2") {
                        action { loadFromMemory() }
                    }
                    item(messages["load_from_sdcard"], "Shortcut+F3") {
                        action { loadFromSdCard() }
                    }
                    separator()
                    item(messages["load_system_vdf"]) {
                        disableWhen { modelProperty.isNull }
                        action { loadSystemVDF() }
                    }
                    item(messages["load_user_vdf"]) {
                        disableWhen { modelProperty.isNull }
                        action { loadUserVDF() }
                    }
                    separator()
                    item(messages["save"], "Shortcut+S") {
                        disableWhen { modelProperty.isNull }
                        action { saveFile() }
                    }
                    separator()
                    item(messages["refresh"], "F5") {
                        disableWhen { modelProperty.isNull }
                        action { refreshModel() }
                    }
                    separator()
                    item(messages["exit"], "Alt+F4") {
                        action { exit() }
                    }
                }

                contextmenu {
                    this@Controller.contextMenu = this
                    item(messages["load_form_file"]) { action { loadFromFile() } }
                    item(messages["load_from_memory"]) { action { loadFromMemory() } }
                    item(messages["load_from_sdcard"]) { action { loadFromSdCard() } }
                    separator()
                    item(messages["load_system_vdf"]) {
                        disableWhen { modelProperty.isNull }
                        action { loadSystemVDF() }
                    }
                    item(messages["load_user_vdf"]) {
                        disableWhen { modelProperty.isNull }
                        action { loadUserVDF() }
                    }
                    separator()
                    item(messages["save"]) {
                        disableWhen { modelProperty.isNull }
                        action { saveFile() }
                    }
                    separator()
                    item(messages["refresh"]) {
                        disableWhen { modelProperty.isNull }
                        action { refreshModel() }
                    }
                    separator()
                    item(messages["exit"]) { action { exit() } }
                }
            }
        }

        center {
            webView = webview {
                setOnMouseClicked { mouseClicked(it) }
                isContextMenuEnabled = false
            }
        }

    }

    private fun disableUI(readOnlyBooleanProperty: ReadOnlyBooleanProperty) {
        primaryStage.scene.cursorProperty().bind(Bindings.`when`(readOnlyBooleanProperty).then(Cursor.WAIT).otherwise(Cursor.DEFAULT))
        root.disableProperty().bind(readOnlyBooleanProperty)
    }

    init {
        log.info(this.javaClass.name)

        val params = app.parameters.unnamed
        if (params != null && params.size > 0 && params[0].endsWith(".mdl")) {
            Platform.runLater { load(File(params[0])) }
        }

        primaryStage.icons.add(Image(javaClass.getResource("/icon.png").toString()))
        primaryStage.title = messages["title"]// TODO, System.getProperty(MdlViewer.version))
    }

    private fun exit() {
        Platform.exit()
    }

    private fun load(file: File) {
        preferences { put(LAST_LOAD_DIR, file.parentFile.absolutePath) }
        model = loadModel(file)
        refreshModel()
    }

    private fun loadFromFile() {
        FileChooser().apply {
            title = messages["load_from_file"]
            preferences {
                val dir = File(get(LAST_LOAD_DIR, MdlViewer.programDir))
                if (dir.exists() && dir.isDirectory) initialDirectory = dir
            }
            extensionFilters.add(ExtensionFilter(messages["mdl_file"], "*.mdl"))
        }.showOpenDialog(primaryStage)?.apply {
            load(this)
        }
    }

    private fun loadFromMemory() {
        selectFromMemory.openDialog()
        selectFromMemory.getResult()?.apply {
            disableUI(runningProperty())
            success {
                modelProperty.set(value)
                refreshModel()
            }
        }
    }

    private fun loadFromSdCard() {
        selectFromSDCard.openDialog()
        selectFromSDCard.getResult()?.apply {
            disableUI(runningProperty())
            success {
                modelProperty.set(value)
                refreshModel()
            }
        }
    }

    private fun loadSystemVDF() {
        FileChooser().apply {
            title = messages["load_system_vdf"]
            preferences {
                val dir = File(get(LAST_SAVE_DIR, MdlViewer.programDir))
                if (dir.exists() && dir.isDirectory) initialDirectory = dir
            }
            extensionFilters.add(ExtensionFilter(messages["vdf_file"], "*.vdf"))
        }.showOpenDialog(primaryStage)?.apply {
            preferences { put(LAST_LOAD_DIR, parentFile.absolutePath) }
            val vdfFile = HoTTDecoder.decodeVDF(this)
            val names = vdfFile.voiceList.map { it.name }.toList()
            Announcements.saveSystemPrefs(model!!.model.transmitterType, names)
            refreshModel()
        }
    }

    private fun loadUserVDF() {
        FileChooser().apply {
            title = messages["load_user_vdf"]
            preferences {
                val dir = File(get(LAST_SAVE_DIR, MdlViewer.programDir))
                if (dir.exists() && dir.isDirectory) initialDirectory = dir
            }
            extensionFilters.add(ExtensionFilter(messages["vdf_file"], "*.vdf"))
        }.showOpenDialog(primaryStage)?.apply {
            preferences { put(LAST_LOAD_DIR, parentFile.absolutePath) }
            val vdfFile = HoTTDecoder.decodeVDF(this)
            val names = vdfFile.voiceList.map { it.name }.toList()
            Announcements.saveUserPrefs(model!!.model.transmitterType, names)
            refreshModel()
        }
    }

    private fun mouseClicked(e: MouseEvent) {
        if (e.button == MouseButton.SECONDARY)
            contextMenu.show(webView, e.screenX, e.screenY)
        else
            contextMenu.hide()
    }

    private fun refreshModel() {
        val m = model
        if (m != null) {
            val tempFile = File.createTempFile("Model", ".html")
            tempFile.deleteOnExit()

            val task = webView.runAsyncWithOverlay {
                FileWriter(tempFile).use { it.write(m.html) }
            } ui {
                webView.engine.load("file:/${tempFile.absolutePath}")
                runLater { tempFile.delete() }
            }

            disableUI(task.runningProperty())
        }
    }

    private fun saveFile() {
        FileChooser().apply {
            title = messages["save"]
            preferences {
                val dir = File(get(LAST_SAVE_DIR, MdlViewer.programDir))
                if (dir.exists() && dir.isDirectory) initialDirectory = dir
            }
            initialFileName = model?.fileName
            extensionFilters.add(ExtensionFilter(messages["html_file"], "*.html"))
            extensionFilters.add(ExtensionFilter(messages["pdf_file"], "*.pdf"))
            extensionFilters.add(ExtensionFilter(messages["xml_file"], "*.xml"))
            extensionFilters.add(ExtensionFilter(messages["mdl_file"], "*.mdl"))
            extensionFilters.add(ExtensionFilter(messages["txt_file"], "*.txt"))
        }.showSaveDialog(primaryStage)?.apply {
            preferences { put(LAST_SAVE_DIR, parentFile.absolutePath) }
            val extension = FilenameUtils.getExtension(name).toLowerCase()

            async {
                when (extension) {
                    "html" -> model?.saveHtml(this@apply)

                    "pdf" -> model?.savePdf(this@apply)

                    "xml" -> model?.saveXml(this@apply)

                    "mdl" -> model?.saveMdl(this@apply)

                    "txt" -> model?.saveTxt(this@apply)

                    else -> {
                    }
                }
            }
        }
    }
}
